check/checksum.go

```diff
@@ -0,0 +1,182 @@
+package check
+
+import (
+	"bytes"
+	"crypto/md5"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/schollz/progressbar/v3"
+)
+
+func isHiddenFile(f string) bool {
+	return string(f[0]) == "."
+}
+
+func checksumFor(path string) ([]byte, error) {
+	h := md5.New()
+
+	var r *os.File
+	r, err := os.Open(path)
+	if err != nil {
+		return nil, fmt.Errorf("error opening file in %s: %w", path, err)
+	}
+	defer r.Close()
+
+	if _, err := io.Copy(h, r); err != nil {
+		return nil, fmt.Errorf("error reading %s content: %w", path, err)
+	}
+
+	return h.Sum(nil), nil
+}
+
+func checkChecksum(src, target string) (bool, error) {
+	name := filepath.Base(src)
+	sc, err := checksumFor(src)
+	if err != nil {
+		return false, fmt.Errorf("error getting %s checksum: %w", src, err)
+	}
+
+	targetPath := filepath.Join(target, name)
+	tc, err := checksumFor(targetPath)
+	if err != nil {
+		return false, fmt.Errorf("error getting %s checksum: %w", targetPath, err)
+	}
+
+	return bytes.Equal(sc, tc), nil
+}
+
+// Check compares md5 checksum files from src directory with corresponding files in target directory
+func CheckChecksum(src, target string) error {
+	ls, err := filepath.Glob(filepath.Join(src, "*.md5"))
+	if err != nil {
+		return fmt.Errorf("error listing checksum files in %s directory: %w", src, err)
+	}
+
+	if len(ls) == 0 {
+		return fmt.Errorf("target directory %s has no checksum files to compare with", target)
+	}
+
+	bar := progressbar.Default(int64(len(ls)))
+	defer bar.Close()
+	bar.Describe("Checking files checksum")
+	if err := bar.RenderBlank(); err != nil {
+		return fmt.Errorf("error rendering the progress bar: %w", err)
+	}
+
+	type qm struct {
+		f     string
+		equal bool
+		err   error
+	}
+	q := make(chan qm, len(ls))
+	defer close(q)
+
+	var wg sync.WaitGroup
+
+	for _, f := range ls {
+		wg.Add(1)
+
+		go func(f string, wg *sync.WaitGroup, q chan<- qm) {
+			defer wg.Done()
+
+			equal, err := checkChecksum(f, target)
+
+			q <- qm{f, equal, err}
+		}(f, &wg, q)
+	}
+
+	wg.Wait()
```

Comment:
A gente não precisa parar a execuão aqui, a gente pode continuar. Mas precisamos do `wg.Wait()` para saber quando fechar o canal e encerrar o `for` que lê o canal. O que acha dessa sugestão?

```suggestion
	go func() {
		wg.Wait()
		close(q)
	}()
```

E removemos o `defer close(q)` da linha 77.

Dessa forma não esperamos todas as gorotinas terminarem para começar a ler o canal…

Reply:
É recomendado pelo go passar o ponteiro do `WaitGroup` quando alterando ele de go rotinas, adicionei isso como parâmetro da função.